### PR TITLE
Docs: polish `sky.Task` doc strings.

### PR DESCRIPTION
**Pull request by concretevitamin**
Changes
- Expose a subset of methods of `sky.Task` to docs
  - Excluded:
    - Experimental/internal-facing: `:exclude-members: estimate_runtime, get_inputs_cloud, set_time_estimator, sync_storage_mounts`
    - Some undocumented methods
    - "Hidden" methods `__rshift__, __repr__`

- CLI docs: tweak order.

Tested:
- rendered locally

### CODE REVIEW

File: sky/task.py [hunk 1]
```diff
@@ -207,7 +226,16 @@ def _validate(self):
                         f'a symlink to a directory). {self.workdir} not found.')
 
     @staticmethod
-    def from_yaml(yaml_path):
+    def from_yaml(yaml_path: str) -> 'Task':
+        """Initializes a task from a task YAML.
+
```
Comment (Michaelvll) [inline sky/task.py:231]: nit: Can we add an example here:
```
Example

    task = sky.Task.from_yaml('/path/to/task.yaml')
```

Reason: one of our user created it with `sky.Task().from_yaml('task.yaml')`, which creates an unnecessary `Task` object

File: sky/task.py [hunk 2]
```diff
@@ -695,6 +754,47 @@ def get_cloud_to_remote_file_mounts(self) -> Optional[Dict[str, str]]:
                 d[k] = v
         return d
 
+    def to_yaml_config(self) -> Dict[str, Any]:
```
Comment (Michaelvll) [inline sky/task.py:757]: Do we need to expose this to the user?

File: sky/task.py [hunk 3]
```diff
@@ -581,94 +724,10 @@ def sync_storage_mounts(self) -> None:
...
-            the VM on which this Task will eventually run on, and local is the
-            node from which the task is launched.
-        """
-        if self.file_mounts is None:
-            self.file_mounts = {}
-        self.file_mounts.update(file_mounts)
-        # For validation logic:
-        return self.set_file_mounts(self.file_mounts)
-
     def get_local_to_remote_file_mounts(self) -> Optional[Dict[str, str]]:
```
Comment (Michaelvll) [inline sky/task.py:727]: Do we need to expose this? Also, for `get_cloud_to_remote_file_mounts` and `to_yaml_config`

**Review by Michaelvll — COMMENTED** [review]
Thank you for adding the documentation of the `Task` class @concretevitamin ! It will be very useful for our API users. Left several comments. : )

A future TODO: we need to make the `Task` class immutable (as what we did for the `Resources`) and ask the user to use the `setter` functions. Otherwise, it is weird to have two ways to change the variables.

File: sky/task.py [hunk 4]
```diff
@@ -86,36 +86,55 @@ def __init__(
...
+
+        All setters of this class, ``Task.set_*()``, return ``self``, i.e.,
+        they are fluent APIs and can be chained together.
+
+        Example:
+            >>> # A Task that will sync up local workdir '.', containing
+            >>> # requirements.txt and train.py.
+            >>> sky.Task(setup='pip install requirements.txt',
+            >>>          run='python train.py',
+            >>>          workdir='.')
```
[the remark this answers is not in the document]
Comment (concretevitamin) [reply sky/task.py:105]: On my local sphinx, `...` doesn't render correctly, with or without `.. code-block:: python`.

File: sky/task.py [hunk 5]
```diff
@@ -344,11 +331,20 @@ def num_nodes(self) -> int:
     def envs(self) -> Dict[str, str]:
         return self._envs
 
-    def set_envs(self, envs: Union[None, Tuple[Tuple[str, str]], Dict[str,
-                                                                      str]]):
+    def set_envs(
+            self, envs: Union[None, Tuple[Tuple[str, str]],
+                              Dict[str, str]]) -> 'Task':
+        """Sets the environment variables for use inside the setup/run commands.
+
```
[the remark this answers is not in the document]
Comment (concretevitamin) [reply sky/task.py:342]: Done, with rewording to reflect the type:
```
        Args:
          envs: (optional) either a list of ``(env_name, value)`` or a dict
            ``{env_name: value}``.
```

File: sky/task.py (hunk 3 above)
Comment (concretevitamin) [reply sky/task.py:737]: Good call, agreed it's a good idea to minimize the advertised API surface. All hidden now.

File: sky/task.py (hunk 2 above)
Comment (concretevitamin) [reply sky/task.py:771]: Good call, agreed it's a good idea to minimize the advertised API surface. All hidden now.

File: sky/task.py [hunk 6]
```diff
@@ -435,39 +440,155 @@ def set_resources(self, resources: Union['resources_lib.Resources',
...
+        (setting ``source``), or backed by an existing cloud bucket (setting
+        ``name`` to the bucket name; or setting ``source`` to the bucket URI).
 
         Example:
-            task.set_storage_mounts({
-                '/tmp/imagenet/': \
-                Storage(name='imagenet', source='s3://imagenet-bucket'):
-            })
+            >>> task.set_storage_mounts({
+            >>>     '/tmp/imagenet/': sky.Storage(source='/data/imagenet'),
```
[the remark this answers is not in the document]
Comment (concretevitamin) [reply sky/task.py]: RE `...`: it doesn't render correctly for me?

RE question: Great catch! Fixed.

File: sky/task.py [hunk 7]
```diff
@@ -435,39 +440,155 @@ def set_resources(self, resources: Union['resources_lib.Resources',
...
+        existing file_mounts (calls ``dict.update()``), rather than
+        overwritting it.
+
+        This should be called before provisioning in order to take effect.
+
+        Example:
+
+            >>> task.update_file_mounts({
+            >>>     '~/.config': '~/Documents/config',
+            >>>     '/tmp/workdir': '/local/workdir/cnn-cifar10',
```
[the remark this answers is not in the document]
Comment (concretevitamin) [reply sky/task.py:552]: it doesn't render correctly for me?

File: sky/task.py [hunk 8]
```diff
@@ -435,39 +440,155 @@ def set_resources(self, resources: Union['resources_lib.Resources',
...
+        paths on the remote VM(s) where this Task will run.
+
+        Neither source or destimation paths can end with a slash.
+
+        Example:
+
+            >>> task.set_file_mounts({
+            >>>     '~/.dotfile': '/local/.dotfile',
+            >>>     # /remote/dir/ will contain the contents of /local/dir/.
+            >>>     '/remote/dir': '/local/dir',
```
[the remark this answers is not in the document]
Comment (concretevitamin) [reply sky/task.py:487]: it doesn't render correctly for me?

File: sky/task.py [hunk 9]
```diff
@@ -435,39 +440,155 @@ def set_resources(self, resources: Union['resources_lib.Resources',
     def get_resources(self):
         return self.resources
 
-    def set_time_estimator(self, func):
-        """Sets a func mapping resources to estimated time (secs)."""
+    def set_time_estimator(self, func) -> 'Task':
+        """Sets a func mapping resources to estimated time (secs).
+
+        This is EXPERIMENTAL.
```
[the remark this answers is not in the document]
Comment (concretevitamin) [reply sky/task.py:455]: Unfortunately, that made the line go over column limit.

File: sky/task.py [hunk 10]
```diff
@@ -435,39 +440,155 @@ def set_resources(self, resources: Union['resources_lib.Resources',
     def get_resources(self):
         return self.resources
 
-    def set_time_estimator(self, func):
-        """Sets a func mapping resources to estimated time (secs)."""
+    def set_time_estimator(self, func) -> 'Task':
```
[the remark this answers is not in the document]
Comment (concretevitamin) [reply sky/task.py]: Done.

File: sky/task.py [hunk 11]
```diff
@@ -418,14 +415,22 @@ def get_outputs(self):
...
         """Sets the required resources to execute this task.
 
+        If this function is not called for a Task, default resource
+        requirements will be used (8 vCPUs).
+
         Args:
           resources: either a sky.Resources, or a set of them.  The latter case
-            indicates the user intent "pick any one of these resources" to run
-            a task.
+            is EXPERIMENTAL and indicates asking the optimizer to "pick any one
```
[the remark this answers is not in the document]
Comment (concretevitamin) [reply sky/task.py]: I picked upper case to serve as a warning (this is common).

RE the latter: Done.

File: sky/task.py (hunk 1 above)
Comment (concretevitamin) [reply sky/task.py:231]: Done, good call.

**Review by concretevitamin — COMMENTED** [review]
PTAL

File: sky/task.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/task.py:105]: Interesting, the reason I mentioned this is in the doc of the `sky.launch`, we have the `...` for continued lines. Is it possible to keep them consistent?

[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/task.py:105]: Seems the problem for the rendering issue is that `...` with `>>>` does not support empty lines. The following works for me:
```
        Example:
            >>> # A Task that will sync up local workdir '.', containing
            >>> # requirements.txt and train.py.
            >>> sky.Task(
            ...     setup='pip install requirements.txt',
            ...     run='python train.py',
            ...     workdir='.',
            ... )
            >>> # An empty Task for provisioning a cluster.
            >>> task = sky.Task(num_nodes=n).set_resources(...)
            >>> # Chaining setters.
            >>> sky.Task().set_resources(...).set_file_mounts(...)
```

**Comment by concretevitamin** [comment]
Reformatted the `Example` code blocks with `.. code-block:: python` and removed the `>>>` prefix. PTAL.

**Review by Michaelvll — APPROVED** [review]
Thanks for fixing it @concretevitamin! LGTM.

**Pull request merged**